runtime/src/accounts_db.rs

```diff
@@ -2881,52 +2938,68 @@ impl AccountsDB {
-                        };
-                        let entry = accum
-                            .entry(stored_account.meta.pubkey)
-                            .or_insert_with(Vec::new);
-                        entry.push((stored_account.meta.write_version, account_info));
-                    },
-                );
-
-            let mut accounts_map: HashMap<Pubkey, Vec<(u64, AccountInfo)>> = HashMap::new();
-            for accumulator_entry in accumulator.iter() {
-                for (pubkey, storage_entry) in accumulator_entry {
-                    let entry = accounts_map.entry(*pubkey).or_insert_with(Vec::new);
-                    entry.extend(storage_entry.iter().cloned());
-                }
-            }
-
+            let accounts_map: AccountsMap = DashMap::new();
+            let storage_maps: Vec<Arc<AccountStorageEntry>> = self
+                .storage
+                .get_slot_stores(*slot)
+                .map(|res| res.read().unwrap().values().cloned().collect())
+                .unwrap_or_default();
+            self.thread_pool.install(|| {
+                storage_maps.par_iter().for_each(|storage| {
+                    let accounts = storage.accounts.accounts(0);
+                    accounts.into_iter().for_each(|stored_account| {
+                        let entry = accounts_map
+                            .get(&stored_account.meta.pubkey)
+                            .unwrap_or_else(|| {
+                                accounts_map
+                                    .entry(stored_account.meta.pubkey)
+                                    .or_insert(Mutex::new(BTreeMap::new()))
+                                    .downgrade()
+                            });
+                        assert!(
+                            // There should only be one update per write version for a specific slot
+                            // and account
+                            entry
+                                .lock()
+                                .unwrap()
+                                .insert(
+                                    stored_account.meta.write_version,
+                                    (storage.append_vec_id(), stored_account)
+                                )
+                                .is_none()
+                        );
+                    })
+                });
+            });
             // Need to restore indexes even with older write versions which may
             // be shielding other accounts. When they are then purged, the
             // original non-shielded account value will be visible when the account
             // is restored from the append-vec
-            if !accumulator.is_empty() {
+            if !accounts_map.is_empty() {
                 let mut _reclaims: Vec<(u64, AccountInfo)> = vec![];
-                for (pubkey, account_infos) in accounts_map.iter_mut() {
-                    account_infos.sort_by(|a, b| a.0.cmp(&b.0));
```

Comment:
nice rewrite; I see the reason this `.sort` is alternatively accomplished just by using `BTreeMap`. :)